versioned_docs/version-4.0.0-alpha.1/datastructures/storage-layout.md

```diff
@@ -0,0 +1,141 @@
+`Non-Packed`, preventing it from being eagerly loaded during arbitrary storage operations.
+
+The following example demonstrates how we can solve the problem introduced in the above 
+section. You'll notice that for the lazily loaded storage field, we now work with getters 
+and setters to access and modify the underlying storage value:
+
+```rust
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[ink::contract]
+mod mycontract {
+    use ink::prelude::vec::Vec;
+    use ink::storage::Lazy;
+
+    #[derive(Default)]
+    #[ink(storage)]
+    pub struct MyContract {
+        tiny_value: Balance,
+        /// This vector might get large and expensive to work with.
+        /// We want to enforce a `Non-Packed` storage layout.
+        large_vec: Lazy<Vec<Balance>>,
+    }
+
+    impl MyContract {
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Self::default()
+        }
+
+        /// Because `large_vec` is loaded lazily, this message is always cheap.
+        #[ink(message)]
+        pub fn get_balance(&self) -> Balance {
+            self.tiny_value
+        }
+
+        /// Lazy fields like `large_vec` provide `get()` and `set()` storage operators.
+        #[ink(message)]
+        pub fn add_balance(&mut self, value: Balance) {
+            let mut balances = self.large_vec.get_or_default();
+            balances.push(value);
+            self.large_vec.set(&balances);
+        }
+    }
+}
+```
+
+:::caution
+
+`Vec` are always loaded as a whole, meaning that all elements of the `Vec` live under a 
+single storage key. Wrapping the `Vec` inside `Lazy`, like the provided example above does, 
+has no influence on its elements. If you are dealing with large or sparse arrays on 
+contract storage, consider using a `Mapping` instead.
+
+:::
+
+## Manual vs. Automatic Key generation
+
+Per default, keys are calculated automatically for you, thanks to the 
+[`AutoKey`](https://docs.rs/ink_storage_traits/4.0.0-beta.1/ink_storage_traits/struct.AutoKey.html)
+primitive. They'll be generated at compile time and ruled out for conflicts. However, for `Non-Packed` types like `Lazy` or the `Mapping`, the 
+[`ManualKey`](https://docs.rs/ink_storage_traits/4.0.0-beta.1/ink_storage_traits/struct.ManualKey.html)
+primitive allows manual control over the storage key of a field like so:
+
+```rust
+#[ink(storage)]
+pub struct MyContract {
+    /// The storage key for this field is always `0x0000007f`
+    inner: Lazy<bool, ManualKey<127>>,
+}
+```
+
+This may be advantegous: Your storage key will always stay the same, regardless of 
+the version of your contract or ink! itself (note that the key calcualtion algorithm may 
+change with future ink! versions).
+
+:::caution
+
+Using `ManualKey` instead of `AutoKey` might be especially desireable for upgradable 
+contracts, as using `AutoKey` might result in a different storage key for the same field
+in a newer version of the contract. Which will break the contract after an upgrade!
```

Comment:
Just took a look and we actually don't use `ManualKey` in our `ink/examples/upgradeable-contracts/` yet.